cloud/run/workflow.py

```diff
@@ -0,0 +1,182 @@
+        with self._shutdownLock:
+            self._shutdown = True
+
+
+ProcessingHook = Callable[[Path, str], str]
+
+
+def execute(
+    paths: Paths,
+    preprocess_hook: ProcessingHook,
+    postprocess_hook: ProcessingHook,
+    bqms_request: CreateMigrationWorkflowRequest,
+    macro_processor: Optional[MacroProcessor] = None,
+    executor_factory: Type[Executor] = SynchronousExecutor,
+) -> None:
+    """Executes the preprocess->translate->postprocess workflow.
+
+    Example:
+        .. code-block::
+
+            from run.gcp.bqms.request import build as build_bqms_request
+            from run.hooks import (
+                postprocess as postprocess_hook,
+                preprocess as preprocess_hook
+            )
+            from run.macro_processor import MacroProcessor
+            from run.paths import Paths
+
+            paths = Paths.from_mapping(...)
+            bqms_request = build_bqms_request(...)
+            macro_processor = MacroProcessor.from_mapping(...)
+            execute(paths, preprocess_hook, postprocess_hook, bqms_request,
+                    macro_processor)
+
+    Args:
+        paths: A run.paths.Paths instance containing project paths.
+        preprocess_hook: A callable for hooking user-defined preprocessing
+            logic into the translation workflow.
+        postprocess_hook: A callable for hooking user-defined postprocessing
+            logic into the translation workflow.
+        bqms_request: A
+            google.cloud.bigquery_migration_v2.CreateMigrationWorkflowRequest
+            for requesting the execution of the BQMS migration workflow.
+        macro_processor: An optional run.macro_processor.MacroProcessor for
+            handling macro/templating languages embedded in code to be
+            translated.
+        executor_factory: An optional concurrent.futures.Executor factory used
+            to create an executor for per-path pre and postprocessing tasks.
+    """
+
+    def _preprocess(source_file_path: Path) -> None:
+        """Preprocessing task.
+
+        To be submitted to executor for each input path.
+        """
+        relative_file_path = source_file_path.relative_to(paths.input_path)
+        target_file_path = paths.preprocessed_path / relative_file_path
+        target_file_path.parent.mkdir(parents=True, exist_ok=True)
+
+        # Do not preprocess metadata zip file.
+        if relative_file_path.name.endswith(".zip"):
+            with source_file_path.open(mode="rb") as source_file:
+                source_bytes = source_file.read()
+            with target_file_path.open(mode="wb") as target_file:
+                target_file.write(source_bytes)
+            return
+
+        with source_file_path.open(mode="r", encoding="utf-8") as source_file:
+            source_text = source_file.read()
+
+        preprocessed_text = preprocess_hook(relative_file_path, source_text)
```

Comment:
So providing a do-nothing preprocess_hook here is indented? We will expect the customer to only look into the hooks.py file only?
This LGTM.  Maybe reflect this with some documentation in the user guide?

Reply:
Yes you are correct. This will definitely be documented in the user guide.